File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/src/LIVE_short_trader_multi_filter/data_client.py
from __future__ import annotations

from datetime import datetime
import time
from typing import Optional

import pandas as pd
import requests

from .config import TraderConfig


class DataClient:
    def __init__(self, config: TraderConfig):
        self.config = config

    def fetch_bybit_bars(
        self,
        symbol: Optional[str] = None,
        category: Optional[str] = None,
        days: Optional[int] = None,
        interval_minutes: Optional[int] = None,
        max_retries: int = 5,
        backoff_seconds: float = 1.5,
    ) -> pd.DataFrame:
        symbol = symbol or self.config.symbol
        category = category or self.config.category
        days = days or self.config.backtest_days
        interval_minutes = interval_minutes or self.config.agg_minutes

        end = int(datetime.utcnow().timestamp())
        start = end - days * 24 * 60 * 60
        df_list = []

        while start < end:
            url = "https://api.bybit.com/v5/market/kline"
            params = {
                "category": category,
                "symbol": symbol,
                "interval": str(interval_minutes),
                "start": start * 1000,
                "limit": 1000,
            }
            attempt = 0
            while True:
                resp = requests.get(url, params=params, timeout=10)
                if not resp.ok:
                    raise RuntimeError(f"Bybit API request failed with status {resp.status_code}: {resp.text}")

                payload = resp.json()
                ret_code = str(payload.get("retCode"))
                if ret_code == "0":
                    break
                if ret_code == "10006" and attempt < max_retries:
                    attempt += 1
                    sleep_for = backoff_seconds * attempt
                    time.sleep(sleep_for)
                    continue
                raise RuntimeError(f"Bybit API returned error code {payload.get('retCode')}: {payload.get('retMsg')}")

            rows = payload["result"].get("list", [])
            if not rows:
                break

            df = pd.DataFrame(rows, columns=["timestamp", "Open", "High", "Low", "Close", "Volume", "turnover"])
            df["timestamp"] = pd.to_datetime(df["timestamp"].astype(int), unit="ms")
            df = df.sort_values("timestamp")
            for col in ["Open", "High", "Low", "Close", "Volume"]:
                df[col] = df[col].astype(float)
            df = df[["timestamp", "Open", "High", "Low", "Close", "Volume"]]
            df.set_index("timestamp", inplace=True)
            df_list.append(df)
            start = int(df.index[-1].timestamp()) + interval_minutes * 60
            time.sleep(0.2)

        if not df_list:
            raise ValueError("No candle data received from Bybit.")

        return pd.concat(df_list).sort_index()

    def fetch_bybit_1m(self, symbol: Optional[str] = None, category: Optional[str] = None, days: Optional[int] = None) -> pd.DataFrame:
        """Backward compatible wrapper for code paths still requesting 1m bars."""
        return self.fetch_bybit_bars(symbol=symbol, category=category, days=days, interval_minutes=1)

File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/Unity App Project/short_trader_multi_filter/Assets/Scripts/Core/DataClient.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;

namespace UnityApp.ShortTraderMultiFilter
{
    public class DataClient
    {
        private static readonly HttpClient Http = new HttpClient();
        private readonly TraderConfig _config;

        public DataClient(TraderConfig config)
        {
            _config = config;
        }

        public List<Candle> FetchBybitBars(
            string? symbol = null,
            string? category = null,
            double? days = null,
            int? intervalMinutes = null,
            int maxRetries = 5,
            double backoffSeconds = 1.5)
        {
            var sym = symbol ?? _config.Symbol;
            var cat = category ?? _config.Category;
            var window = days ?? _config.BacktestDays;
            var interval = intervalMinutes ?? _config.AggregationMinutes;

            var end = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            var start = end - (long)(window * 24 * 60 * 60);
            var candles = new List<Candle>();

            while (start < end)
            {
                var url = $"https://api.bybit.com/v5/market/kline?category={cat}&symbol={sym}&interval={interval}&start={start * 1000}&limit=1000";
                var attempt = 0;
                JsonElement root;

                while (true)
                {
                    var resp = Http.GetAsync(url).GetAwaiter().GetResult();
                    if (!resp.IsSuccessStatusCode)
                    {
                        throw new InvalidOperationException($"Bybit API request failed ({resp.StatusCode}): {resp.ReasonPhrase}");
                    }

                    var content = resp.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                    root = JsonDocument.Parse(content).RootElement;
                    var retCode = root.TryGetProperty("retCode", out var codeEl) ? codeEl.GetInt32() : -1;
                    if (retCode == 0)
                    {
                        break;
                    }

                    if (retCode == 10006 && attempt < maxRetries)
                    {
                        attempt++;
                        Task.Delay(TimeSpan.FromSeconds(backoffSeconds * attempt)).Wait();
                        continue;
                    }

                    var msg = root.TryGetProperty("retMsg", out var msgEl) ? msgEl.GetString() : "unknown error";
                    throw new InvalidOperationException($"Bybit API returned error code {retCode}: {msg}");
                }

                if (!root.TryGetProperty("result", out var result) || !result.TryGetProperty("list", out var rows) || rows.ValueKind != JsonValueKind.Array)
                {
                    break;
                }

                var chunk = new List<Candle>();
                foreach (var row in rows.EnumerateArray())
                {
                    if (row.ValueKind != JsonValueKind.Array || row.GetArrayLength() < 6)
                    {
                        continue;
                    }

                    var timestamp = long.Parse(row[0].GetString() ?? "0", CultureInfo.InvariantCulture);
                    var open = double.Parse(row[1].GetString() ?? "0", CultureInfo.InvariantCulture);
                    var high = double.Parse(row[2].GetString() ?? "0", CultureInfo.InvariantCulture);
                    var low = double.Parse(row[3].GetString() ?? "0", CultureInfo.InvariantCulture);
                    var close = double.Parse(row[4].GetString() ?? "0", CultureInfo.InvariantCulture);
                    var volume = double.Parse(row[5].GetString() ?? "0", CultureInfo.InvariantCulture);

                    var dt = DateTimeOffset.FromUnixTimeMilliseconds(timestamp).UtcDateTime;
                    chunk.Add(new Candle(dt, open, high, low, close, volume));
                }

                if (chunk.Count == 0)
                {
                    break;
                }

                chunk.Sort((a, b) => a.Timestamp.CompareTo(b.Timestamp));
                candles.AddRange(chunk);

                var last = chunk[^1].Timestamp;
                start = new DateTimeOffset(last).ToUnixTimeSeconds() + interval * 60;
            }

            if (candles.Count == 0)
            {
                throw new InvalidOperationException("No candle data received from Bybit.");
            }

            candles.Sort((a, b) => a.Timestamp.CompareTo(b.Timestamp));
            return candles;
        }
    }
}
